packages/ics721/src/execute.rs

```diff
@@ -0,0 +1,473 @@
+    }
+
+    fn execute(
+        &self,
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        msg: ExecuteMsg,
+    ) -> Result<Response<T>, ContractError> {
+        PO.error_if_paused(deps.storage)?;
+        match msg {
+            ExecuteMsg::ReceiveNft(cw721::Cw721ReceiveMsg {
+                sender,
+                token_id,
+                msg,
+            }) => self.execute_receive_nft(deps, env, info, token_id, sender, msg),
+            ExecuteMsg::ReceiveProxyNft { eyeball, msg } => {
+                self.execute_receive_proxy_nft(deps, env, info, eyeball, msg)
+            }
+            ExecuteMsg::Pause {} => self.execute_pause(deps, info),
+            ExecuteMsg::Callback(msg) => self.execute_callback(deps, env, info, msg),
+        }
+    }
+
+    fn execute_receive_nft(
+        &self,
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        token_id: String,
+        sender: String,
+        msg: Binary,
+    ) -> Result<Response<T>, ContractError> {
+        if PROXY.load(deps.storage)?.is_some() {
+            Err(ContractError::Unauthorized {})
+        } else {
+            self.receive_nft(deps, env, info, TokenId::new(token_id), sender, msg)
+        }
+    }
+
+    fn execute_receive_proxy_nft(
+        &self,
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        eyeball: String,
+        msg: cw721::Cw721ReceiveMsg,
+    ) -> Result<Response<T>, ContractError> {
+        if PROXY
+            .load(deps.storage)?
+            .map_or(true, |proxy| info.sender != proxy)
+        {
+            return Err(ContractError::Unauthorized {});
+        }
+        let mut info = info;
+        info.sender = deps.api.addr_validate(&eyeball)?;
+        let cw721::Cw721ReceiveMsg {
+            token_id,
+            sender,
+            msg,
+        } = msg;
+        let res = self.receive_nft(deps, env, info, TokenId::new(token_id), sender, msg)?;
+        // override method key
+        Ok(res.add_attribute("method", "execute_receive_proxy_nft"))
+    }
+
+    fn get_class_data(&self, deps: &DepsMut, sender: &Addr) -> StdResult<ClassData> {
+        // cw721 v0.17 and higher holds ownership in the contract
+        let ownership: StdResult<Ownership<Addr>> = deps
+            .querier
+            .query_wasm_smart(sender, &cw721_base::msg::QueryMsg::Ownership::<Addr> {});
+        let owner = match ownership {
+            Ok(ownership) => ownership.owner.map(|a| a.to_string()),
+            Err(_) => {
+                // cw721 v0.16 and lower holds minter
+                let minter_response: cw721_base_016::msg::MinterResponse = deps
+                    .querier
+                    .query_wasm_smart(sender, &cw721_base_016::QueryMsg::Minter::<Empty> {})?;
+                deps.api.addr_validate(&minter_response.minter)?;
+                Some(minter_response.minter)
```

Comment:
queries class data from collection (sender), please note here, it checks if contract is older version v0.16 or later) and retrieves owner from `minter` storage!

Reply:
Is ownership error the only way to know if you need to use cw721_base_016 or could you know that by matching on something up front? Not sure if we are using error for non-error logic or if that's the only way. ,

Reply:
Actually latest cw721-base contract still has Minter query - which basically queries for ownership. So it would be possible querying for minter  - where it works for old and new contracts. But this should/will be deprecated and most likely will be remobed. So I thouhg it is better querying for ownership. Imo, checking error is fine here - since we won't all versions being able to migrate.